src/spec.c

```diff
@@ -518,7 +557,29 @@ static int parseVectorField_flat(FieldSpec *fs, ArgsCursor *ac, QueryError *stat
     VECSIM_ERR_MANDATORY(status, VECSIM_ALGORITHM_BF, VECSIM_DISTANCE_METRIC);
     return 0;
   }
+  // Calculating expected blob size of a vector in bytes.
   fs->vectorOpts.expBlobSize = fs->vectorOpts.vecSimParams.bfParams.dim * VecSimType_sizeof(fs->vectorOpts.vecSimParams.bfParams.type);
+  size_t elementSize = VecSimIndex_EstimateElementSize(&fs->vectorOpts.vecSimParams);
+  // Calculating max block size (in # of vectors), according to memory limits
+  size_t maxBlockSize = BLOCK_MEMORY_LIMIT / elementSize;
+  // if Block size was not set by user, sets the default to min(maxBlockSize, DEFAULT_BLOCK_SIZE)
+  if (fs->vectorOpts.vecSimParams.bfParams.blockSize == 0) {
+    if (maxBlockSize < DEFAULT_BLOCK_SIZE ) {
+      fs->vectorOpts.vecSimParams.bfParams.blockSize = maxBlockSize;
+    } else {
+      fs->vectorOpts.vecSimParams.bfParams.blockSize = DEFAULT_BLOCK_SIZE;
+    }
+  }
+  // Calculating index size estimation, after first vector block was allocated.
+  size_t index_size_estimation = VecSimIndex_EstimateInitialSize(&fs->vectorOpts.vecSimParams);
+  index_size_estimation += elementSize * fs->vectorOpts.vecSimParams.bfParams.blockSize;
+  if (index_size_estimation > memoryLimit - used_memory) {
+    QueryError_SetErrorFmt(status, QUERY_ELIMIT, "Vector index size exceeded server limit (%zuB)", memoryLimit);
```

Comment:
late wakeup on my side: consider adding for both error messages the expected vs actual memory sizes

Reply:
I thought about it and decided that it a bit silly to print back a number that was specifically given the line before, but on second thought it will probably be useful for client logs